File: 2021/02/main.go
package main

import (
	"fmt"
	"bufio"
	"os"
    "strconv"
	"strings"
)

func get_readings(file *os.File) []string {
	var readings []string

	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		line := scanner.Text()
		readings = append(readings, line)
	}
	if err := scanner.Err(); err != nil {
		fmt.Fprintln(os.Stderr, "reading standard input:", err)
	}
	
	return readings
}

func main() {
	readings := get_readings(os.Stdin)

	// Part 1
	depth := 0
	horizontal_position := 0

	for _, element := range readings {
		s := strings.Split(element, " ")

		direction := s[0]
		distance_moved, _ := strconv.Atoi(s[1])

		if direction == "forward" {
			horizontal_position += distance_moved
		}
		if direction == "up" {
			depth -= distance_moved
		}
		if direction == "down" {
			depth += distance_moved
		}
	}

	fmt.Println("###Part 1###")
	fmt.Println("Depth:", depth)
	fmt.Println("Horizontal position:", horizontal_position)
	fmt.Println("Product:", depth * horizontal_position)

	// Part 2
	depth = 0
	horizontal_position = 0
	aim := 0

	for _, element := range readings {
		s := strings.Split(element, " ")

		direction := s[0]
		amount, _ := strconv.Atoi(s[1])

		if direction == "forward" {
			horizontal_position += amount
			depth += aim * amount
		}
		if direction == "up" {
			aim -= amount
		}
		if direction == "down" {
			aim += amount
		}
	}

	fmt.Println("###Part 2###")
	fmt.Println("Depth:", depth)
	fmt.Println("Horizontal position:", horizontal_position)
	fmt.Println("Product:", depth * horizontal_position)
}

File: 2021/02/python/main.py
import sys


def part_one(commands):
    horizontal_position = 0
    depth = 0
    for command in commands:
        direction = command[0]
        magnitude = int(command[1])
        if direction == "forward":
            horizontal_position += magnitude
        elif direction == "up":
            depth -= magnitude
        elif direction == "down":
            depth += magnitude
    return horizontal_position * depth


def part_two(commands):
    horizontal_position = 0
    depth = 0
    aim = 0
    for command in commands:
        direction = command[0]
        magnitude = int(command[1])
        if direction == "forward":
            horizontal_position += magnitude
            depth += aim * magnitude
        elif direction == "up":
            aim -= magnitude
        elif direction == "down":
            aim += magnitude
    return horizontal_position * depth


def main():
    path_to_file = sys.argv[1]
    with open(path_to_file) as f:
        commands = [tuple(line.strip().split()) for line in f.readlines()]

    print(f"Part One: {part_one(commands)}")
    print(f"Part Two: {part_two(commands)}")


if __name__ == "__main__":
    main()
